### ISSUE TITLE
Add examples for another hooks?

### ISSUE THREAD

**Issue by sosukesuzuki** [piotrwitek/react-redux-typescript-guide#120]
Hi, I added example for `useReducer` with #118 .
What do you think about adding examples of another hooks?

**Comment by piotrwitek** [comment]
Yeah, that's a good idea 👍 I'll need to figure out some clear and concise examples.

- [ ] useState - let's use basic Counter example from react docs:
```ts
function CounterWithState({initialCount}) {
  const [count, setCount] = useState(initialCount);
  return (
    <>
      Count: {count}
      <button onClick={() => setCount(initialCount)}>Reset</button>
      <button onClick={() => setCount(prevCount => prevCount + 1)}>+</button>
      <button onClick={() => setCount(prevCount => prevCount - 1)}>-</button>
    </>
  );
}
```

- [ ] useEffect
```ts
function ClickCounterWithEffect({initialCount}) {
  const [count, setCount] = useState(0);

  useEffect(() => {
    // Update the document title using the browser API
    document.title = `You clicked ${count} times`;
	return () => { document.title = '' };
  }, [count]);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button onClick={() => setCount(count + 1)}>
        Click me
      </button>
    </div>
  );
}
```

- [ ] useRef
```ts
function TextInputWithFocusButton() {
  const inputEl = useRef(null);
  const onButtonClick = () => {
    // `current` points to the mounted text input element
    inputEl.current.focus();
  };
  return (
    <>
      <input ref={inputEl} type="text" />
      <button onClick={onButtonClick}>Focus the input</button>
    </>
  );
}
```

- [ ] useContext - TODO

**Comment by sosukesuzuki** [comment]
@piotrwitek I wrote example for useContext. What do you think about this?
This was rewritten with reference to https://reactjs.org/docs/context.html#updating-context-from-a-nested-component (with hooks).
```tsx
interface Theme {
  foreground: string;
  background: string;
}

interface Themes {
  light: Theme;
  dark: Theme;
}

export const themes: Themes = {
  light: {
    foreground: "#000000",
    background: "#eeeeee"
  },
  dark: {
    foreground: "#ffffff",
    background: "#222222"
  }
};

const ThemeContext = React.createContext({
  theme: themes.dark,
  toggleTheme: () => {}
});

function ThemeTogglerButton() {
  const { theme, toggleTheme } = React.useContext(ThemeContext);
  return (
    <button onClick={toggleTheme} style={{ backgroundColor: theme.background }}>
      Toggle Theme
    </button>
  );
}

function Content() {
  return (
    <div>
      <ThemeTogglerButton />
    </div>
  );
}

export default function App() {
  const [theme, setTheme] = React.useState(themes.light);
  function toggleTheme() {
    theme === themes.dark ? setTheme(themes.light) : setTheme(themes.dark);
  }
  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      <Content />
    </ThemeContext.Provider>
  );
}
```
https://codesandbox.io/s/nko6l15xpl

**Comment by piotrwitek** [comment]
Yeah, looks good.

But in the case of Context I was thinking about two things.
- First add a separate component example with Context, because we need that as well.
Basically this:
```ts
interface Theme {
  foreground: string;
  background: string;
}

interface Themes {
  light: Theme;
  dark: Theme;
}

export const themes: Themes = {
  light: {
    foreground: "#000000",
    background: "#eeeeee"
  },
  dark: {
    foreground: "#ffffff",
    background: "#222222"
  }
};

const ThemeContext = React.createContext({
  theme: themes.dark,
  toggleTheme: () => {}
});
```
- Then reuse above context component for useContext example (will cut a bit on boilerplate code for this example)
```ts
function ThemeTogglerButton() {
  return (
    <button onClick={toggleTheme} style={{ backgroundColor: theme.background }}>
      Toggle Theme
    </button>
  );
}

export default function App() {
  const [theme, setTheme] = React.useState(themes.light);
  function toggleTheme() {
    theme === themes.dark ? setTheme(themes.light) : setTheme(themes.dark);
  }
  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      <ThemeTogglerButton />
    </ThemeContext.Provider>
  );
}
```

**Comment by piotrwitek** [comment]
Sure, what I meant is that the example is too big when presented in a single file form. Just split useContext example into two separate files, as it is laid out in my comment above (btw. I updated the code).

Note that in fact we already have this example that I added long time ago, will need a bit or refactoring: https://github.com/piotrwitek/react-redux-typescript-guide/blob/master/playground/src/context/theme-provider.tsx
